### ISSUE TITLE
fcntl( fd, F_GETSIG ) is needed by Wine

### ISSUE THREAD

**Issue by fracting** [mozilla/rr#1560]
Dear rr developers,

rr is a great project which makes Wine developers' live much more happy, thanks a lot for your work!

Today rr has basic support of Wine, but it is not completely yet. Wine Staging is an official part of Wine, which introduce some new feature hasn't been included to plain Wine yet.

In Wine Staging, F_GETSIG is used but it is not supported by rr yet.

 626 /* returns the pipe flags for a file descriptor */
 627 static inline int get_pipe_flags(int fd)
 628 {
 629 #ifdef __linux__
 630     int flags = fcntl( fd, F_GETSIG );
 631 
 632     /* NAMED_PIPE_MESSAGE_STREAM_READ only allowed in
 633      * combination with NAMED_PIPE_MESSAGE_STREAM_WRITE. */
 634     if (!(flags & NAMED_PIPE_MESSAGE_STREAM_WRITE))
 635         flags &= ~NAMED_PIPE_MESSAGE_STREAM_READ;
 636 
 637     return flags;
 638 #else
 639     return 0;
 640 #endif

https://github.com/wine-compholio/wine-patched/blob/81667786de2e23faae28f4f0f1d4a08209b92e1b/dlls/ntdll/file.c#L630

The below hack makes Wine Staging happier with rr:

--- a/src/record_syscall.cc
+++ b/src/record_syscall.cc
@@ -1713,6 +1713,7 @@ static Switchable rec_prepare_syscall_arch(Task* t,
         case Arch::SETOWN:
         case Arch::SETOWN_EX:
         case Arch::SETSIG:
+        case Arch::GETSIG:
           break;

Could you have a look and add GETSIG support?

Thanks!

**Comment by rocallahan** [comment]
> rr is a great project which makes Wine developers' live much more happy, thanks a lot for your work!

Great! Can you tell me how you're using rr? Any online discussion or blog posts I could look at?

**Comment by fracting** [comment]
Thanks for the commit!

I pull the change and retest with latest git, confirming works for me.

For how I use rr:

I just found rr several hours before, and I found it is exactly the tool I'm looking for in the past few years  ;-)

For using rr with Wine, I need to disable Wine's preloader using this customer hack: https://gist.github.com/fracting/d3e193242fd5f41d2f91

Regarding real world debugging case: no, not now. I'm looking for a reverse debugging framework basically because I try to fix this bug http://permalink.gmane.org/gmane.os.cygwin/155768 for many days but haven't success, after meeting with rr I start to have hope ;-)

I'm currently doing testing with Wine + rr, will keep reporting bugs to rr, and see how far we can go. Later then I'll definitely share my experience and recommend rr to more Wine developers, as well as win32 open source developers.

Thank you!
